log.c

```diff
@@ -0,0 +1,129 @@
+/*! \file     log.c
+ * \author    Jay Ridgeway <jayridge@gmail.com>
+ * \copyright GNU General Public License v3
+ * \brief     Buffered logging
+ * \details   Implementation of a simple buffered logger designed to remove
+ * I/O wait from threads that may be sensitive to such delays. Buffers are
+ * saved and reused to reduce allocation calls.
+ *
+ * \ingroup core
+ * \ref core
+ */
+
+#include "log.h"
+
+#define INITIAL_BUFSZ 2048
+#define THREAD_NAME   "log"
+
+
+static gint     initialized = 0;
+static gint     stopping = 0;
+/* Maximum sleep in ms for the print thread */
+static gint     maxdelay = 3000;
+/* Buffers over this size will be freed */
+static gsize    maxbuffer = 1024*16;
+static GMutex   lock;
+static GCond    cond;
+static GSList   *printqueue = NULL;
+static GQueue   *freebufs = NULL;
+static GThread  *printthread = NULL;
+
+
+static void stringfree(void *s)
+{
+	if (s) {
+		g_string_free((GString *)s, TRUE);
+	}
+}
+
+static GString * getbuf(void)
+{
+	GString *s;
+
+	g_mutex_lock(&lock);
+	s = (GString *)g_queue_pop_head(freebufs);
+	g_mutex_unlock(&lock);
+	if (s == NULL) {
+		s = g_string_sized_new(INITIAL_BUFSZ);
+	}
+	return s;
+}
+
+static void * janus_log_thread(void *ctx)
+{
+	GSList  *head, *p;
+	GString *s;
+	gint64  abstime;
+
+	while (!g_atomic_int_get(&stopping)) {
+		abstime = g_get_monotonic_time() + maxdelay * G_TIME_SPAN_MILLISECOND;
+		g_mutex_lock(&lock);
+		g_cond_wait_until(&cond, &lock, abstime);
+		head = printqueue;
```

Comment:
I think you want to check `printqueue` before you `g_cond_wait...` because the wakeup signal for a message could have come before you `g_cond_wait...`